hbase-balancer/src/main/java/org/apache/hadoop/hbase/master/balancer/StochasticLoadBalancer.java

```diff
@@ -408,34 +433,53 @@ boolean needsBalance(TableName tableName, BalancerClusterState cluster) {
 
   @RestrictedApi(explanation = "Should only be called in tests", link = "",
       allowedOnPath = ".*(/src/test/.*|StochasticLoadBalancer).java")
-  BalanceAction nextAction(BalancerClusterState cluster) {
-    return getRandomGenerator().generate(cluster);
+  Pair<CandidateGenerator, BalanceAction> nextAction(BalancerClusterState cluster) {
+    CandidateGenerator generator = getRandomGenerator(cluster);
+    return Pair.newPair(generator, generator.generate(cluster));
   }
 
   /**
    * Select the candidate generator to use based on the cost of cost functions. The chance of
    * selecting a candidate generator is propotional to the share of cost of all cost functions among
    * all cost functions that benefit from it.
    */
-  protected CandidateGenerator getRandomGenerator() {
+  protected CandidateGenerator getRandomGenerator(BalancerClusterState cluster) {
     double sum = 0;
-    for (int i = 0; i < weightsOfGenerators.length; i++) {
-      sum += weightsOfGenerators[i];
-      weightsOfGenerators[i] = sum;
+    for (Class<? extends CandidateGenerator> clazz : candidateGenerators.keySet()) {
+      sum += weightsOfGenerators.getOrDefault(clazz, 0.0);
     }
     if (sum == 0) {
-      return candidateGenerators.get(0);
+      return candidateGenerators.values().stream().findAny().orElseThrow();
+    }
+
+    for (Class<? extends CandidateGenerator> clazz : candidateGenerators.keySet()) {
+      weightsOfGenerators.put(clazz,
+        Math.min(CandidateGenerator.MAX_WEIGHT, weightsOfGenerators.get(clazz) / sum));
     }
-    for (int i = 0; i < weightsOfGenerators.length; i++) {
-      weightsOfGenerators[i] /= sum;
+
+    for (RegionPlanConditional conditional : balancerConditionals.getConditionals()) {
+      Optional<RegionPlanConditionalCandidateGenerator> generator =
+        conditional.getCandidateGenerator();
+      if (generator.isPresent() && generator.get().getWeight(cluster) > 0) {
+        return generator.get();
+      }
     }
+
     double rand = ThreadLocalRandom.current().nextDouble();
-    for (int i = 0; i < weightsOfGenerators.length; i++) {
-      if (rand <= weightsOfGenerators[i]) {
-        return candidateGenerators.get(i);
+    Class<? extends CandidateGenerator> greatestWeightClazz = null;
+    double greatestWeight = 0;
+    for (Class<? extends CandidateGenerator> clazz : candidateGenerators.keySet()) {
+      double generatorWeight = weightsOfGenerators.get(clazz);
+      if (generatorWeight > greatestWeight) {
+        greatestWeight = generatorWeight;
+        greatestWeightClazz = clazz;
+      }
+      if (rand <= generatorWeight) {
+        return candidateGenerators.get(clazz);
       }
     }
-    return candidateGenerators.get(candidateGenerators.size() - 1);
```

Comment:
Previously, this method would often just return the last candidate generator in the list. This refactored implementation does a few things differently:
1. It prioritizes conditional candidate generators over all others
2. In the non-conditional case, it prioritizes the generator with the highest weight instead of the last one in the (no longer existent) list